validator/keymanager/remote.go

```diff
@@ -0,0 +1,255 @@
+	  "ca_cert": "/home/eth2/certs/ca.crt"         // Certificate file for the CA that signed the server's certificate
+	  "client_cert": "/home/eth2/certs/client.crt" // Certificate file for this client
+	  "client_key": "/home/eth2/certs/client.key"  // Key file for this client
+	}
+  }`
+
+// NewRemote creates a key manager populated with the keys from walletd.
+func NewRemote(input string) (KeyManager, string, error) {
+	opts := &remoteOpts{}
+	err := json.Unmarshal([]byte(input), opts)
+	if err != nil {
+		return nil, remoteOptsHelp, err
+	}
+
+	if len(opts.Accounts) == 0 {
+		return nil, remoteOptsHelp, errors.New("at least one account specifier is required")
+	}
+
+	// Load the client certificates.
+	clientPair, err := tls.LoadX509KeyPair(opts.Certificates.ClientCert, opts.Certificates.ClientKey)
+	if err != nil {
+		return nil, remoteOptsHelp, errors.Wrap(err, "failed to obtain client's certificate and/or key")
+	}
+
+	// Load the CA for the server certificate.
+	serverCA, err := ioutil.ReadFile(opts.Certificates.CACert)
+	if err != nil {
+		return nil, remoteOptsHelp, errors.Wrap(err, "failed to obtain server's CA certificate")
+	}
+	cp := x509.NewCertPool()
+	if !cp.AppendCertsFromPEM(serverCA) {
+		return nil, remoteOptsHelp, errors.Wrap(err, "failed to add server's CA certificate to pool")
+	}
+
+	tlsCfg := &tls.Config{
+		Certificates: []tls.Certificate{clientPair},
+		RootCAs:      cp,
+	}
+	clientCreds := credentials.NewTLS(tlsCfg)
+
+	grpcOpts := []grpc.DialOption{
+		// Require TLS with client certificate.
+		grpc.WithTransportCredentials(clientCreds),
+	}
+
+	conn, err := grpc.Dial(opts.Location, grpcOpts...)
+	if err != nil {
+		return nil, remoteOptsHelp, errors.New("failed to connect to remote wallet")
+	}
+
+	km := &Remote{
+		conn:  conn,
+		paths: opts.Accounts,
+	}
+
+	err = km.RefreshValidatingKeys()
+	if err != nil {
+		return nil, remoteOptsHelp, errors.New("failed to fetch accounts from remote wallet")
+	}
+
+	return km, remoteOptsHelp, nil
+}
+
+// FetchValidatingKeys fetches the list of public keys that should be used to validate with.
+func (km *Remote) FetchValidatingKeys() ([][48]byte, error) {
+	res := make([][48]byte, 0, len(km.accounts))
+	for _, accountInfo := range km.accounts {
+		res = append(res, bytesutil.ToBytes48(accountInfo.PubKey))
+	}
+	return res, nil
+}
+
+// Sign without protection is not supported by remote keymanagers.
+func (km *Remote) Sign(pubKey [48]byte, root [32]byte) (*bls.Signature, error) {
+	return nil, errors.New("remote keymanager does not support unprotected signing")
+}
+
+// SignGeneric signs a generic message for the validator to broadcast.
+func (km *Remote) SignGeneric(pubKey [48]byte, root [32]byte, domain [32]byte) (*bls.Signature, error) {
+	accountInfo, exists := km.accounts[pubKey]
+	if !exists {
+		return nil, ErrNoSuchKey
+	}
+
+	client := pb.NewSignerClient(km.conn)
+	req := &pb.SignRequest{
+		Id:     &pb.SignRequest_Account{Account: accountInfo.Name},
```

Comment:
why identify the account by name and not by the pub key?

Reply:
It's a bit of future proofing for key abstraction in the future.